### PR TITLE
Feat/stats

### PR DESCRIPTION
**Pull request by SpencerAxelrod**
Link to JIRA ticket if there is one: 

Updating #378

[Link to feature doc](https://docs.google.com/document/d/1emYw1laZ-YGN2ixMg7M54RDb45QRfrROVEWOvsTcT1I)

### New Features
- Pre-computed stats table replaces full-table `COUNT(*)/SUM(size)` on GET /_stats
- Historical stats queries: GET /_stats?month=X&year=Y

### Deployment changes

- reconcile_stats.py can be used in the container for stats reconciliation. Could be a Cronjob?

### CODE REVIEW

File: indexd/index/drivers/alchemy.py [hunk 1]
```diff
@@ -1794,6 +1861,34 @@ def delete_bundle(self, bundle_id):
...
+                            StatsRecord.year < int(year),
+                        )
+                    )
+                    .order_by(StatsRecord.year.desc(), StatsRecord.month.desc())
+                    .first()
+                )
+                return (stats.total_record_count, stats.total_record_bytes)
+            except Exception as e:
+                self.logger.warning(f"Failed to get stats: {e}")
+                return (None, None)
```
Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py]: Let's raise an exception instead of returning None. Actually, let's just remove the whole try/catch block, there is no reason we should expect this to fail

**Review by paulineribeyre — DISMISSED** [review]

File: indexd/index/drivers/alchemy.py [hunk 2]
```diff
@@ -1794,6 +1861,34 @@ def delete_bundle(self, bundle_id):
...
+                stats = (
+                    session.query(StatsRecord)
+                    .filter(
+                        or_(
+                            and_(
+                                StatsRecord.month <= int(month),
+                                StatsRecord.year == int(year),
+                            ),
+                            StatsRecord.year < int(year),
+                        )
```
[the remark this answers is not in the document]
Comment (SpencerAxelrod) [reply indexd/index/drivers/alchemy.py]: we need for example: if you query for january 2026, it would exclude december 2025 (which might be the most recent)

File: indexd/index/drivers/alchemy.py (hunk 1 above)
Comment (SpencerAxelrod) [reply indexd/index/drivers/alchemy.py]: Yes that makes sense, I will remove try catch. But we would still need some kind of none check and return (0, 0), for legitimate historical queries from before the first row right? Like if first record is 2025, and they query for 2020?

Comment (paulineribeyre) [reply indexd/index/drivers/alchemy.py]: True

File: indexd/index/drivers/alchemy.py [hunk 3]
```diff
@@ -382,6 +376,44 @@ def update_stats(session, number, size=None):
...
+                    StatsRecord.year == int(year),
+                ),
+                StatsRecord.year < int(year),
+            )
+        )
+        .order_by(StatsRecord.year.desc(), StatsRecord.month.desc())
+        .first()
+    )
+    if stats is None:
+        return (0, 0)
```
Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py:422]: could you add a quick comment to explain when we might get this edge case

Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py]: and unit test

File: indexd/index/drivers/alchemy.py [hunk 4]
```diff
@@ -783,6 +870,7 @@ def add(
 
                 if self.config.get("ADD_PREFIX_ALIAS"):
                     self.add_prefix_alias(record, session)
+                update_stats(session, 1, size or 0)
```
Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py]: oh now i understand why you were doing `if size is None: size = 0` in `update_stats` 🤦‍♀️ Sorry, the previous version was better, can you put it back instead of passing `size or 0` as argument

File: indexd/stats_table_migration.py [hunk 5]
```diff
@@ -7,32 +7,29 @@
...
+import sqlalchemy as sa
 from cdislogging import get_logger
 from sqlalchemy import and_, func
 
+from indexd.index.drivers.alchemy import IndexRecord, StatsRecord
+
 logger = get_logger(__name__)
 
 
 def seed_stats_from_connection(bind):
```
Comment (paulineribeyre) [inline indexd/stats_table_migration.py:19]: The file could be renamed `stats_utils.py` or the functions could be moved to `indexd/utils.py`

File: indexd/index/drivers/alchemy.py (hunk 1 above)
Comment (paulineribeyre) [reply indexd/index/drivers/alchemy.py]: Could you add a unit test for what you mentioned above:
> we would still need some kind of none check and return (0, 0), for legitimate historical queries from before the first row right? Like if first record is 2025, and they query for 2020?

File: indexd/stats_table_migration.py [hunk 6]
```diff
@@ -1,93 +0,0 @@
-"""
```
Comment (paulineribeyre) [inline indexd/stats_table_migration.py]: you deleted this file but did not commit `stats_utils.py`

File: migrations/versions/9a2169051163_createstatstable.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        ),
+        sa.Column(
+            "total_record_bytes", sa.BIGINT(), autoincrement=False, nullable=False
+        ),
+        sa.Column("month", sa.INTEGER(), autoincrement=False, nullable=False),
+        sa.Column("year", sa.INTEGER(), autoincrement=False, nullable=False),
+    )
+    op.create_index("ix_stats_year_month", "stats", ["year", "month"])
+
+    from indexd.stats_table_migration import seed_stats_from_connection
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply migrations/versions/9a2169051163_createstatstable.py]: ⬆️

File: indexd/index/drivers/alchemy.py (hunk 1 above)
Comment (paulineribeyre) [reply indexd/index/drivers/alchemy.py]: ⬆️

File: indexd/index/drivers/alchemy.py [hunk 8]
```diff
@@ -327,6 +339,89 @@ def get_record_if_exists(did, session):
...
+                    StatsRecord.year == int(year),
+                ),
+                StatsRecord.year < int(year),
+            )
+        )
+        .order_by(StatsRecord.year.desc(), StatsRecord.month.desc())
+        .first()
+    )
+    if stats is None:
+        # This is the case where stats row exists.
```
Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py]:
```suggestion
        # This is the case where no stats row exists.
```

File: indexd/index/drivers/alchemy.py [hunk 9]
```diff
@@ -327,6 +339,89 @@ def get_record_if_exists(did, session):
...
+        # NOTE: The Alembic migration seeds the stats table automatically on first deploy,
+        # so this should be rare (e.g. a manual intervention).
+        # We COULD fall back to a COUNT(*)/SUM(size) on index_record (expensive)
+        # In this case fix the root cause by running the reconciliation command,
+        # which recomputes stats from index_record and backfills the stats table:
+        #   python bin/reconcile_stats.py
+        return (0, 0)
```
Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py:422]: isn't this ⬇️  the edge case?
> legitimate historical queries from before the first row right? Like if first record is 2025, and they query for 2020?

**Review by paulineribeyre — APPROVED** [review]
